### PR TITLE
Added support for build key in v3

### LINKED ISSUE #636: Add `build` key support for v3 of Docker Compose
This will have to be fixed upstream as the code here in `docker/cli` ignores build at the moment: https://github.com/docker/cli/blob/master/cli/compose/types/types.go#L9

This is a tracking issue as #600 has yet to be merged.

### PR DESCRIPTION
**Pull request by surajnarwade**
Resolves #636 

This PR will add support for `build` in docker compose v3, as docker/cli#481 got merged now.
also this will update `sirupsen`, `docker/cli`, `docker/libcompose` in `glide`,
Also changed `Sirupsen` with `sirupsen` in all kompose packages as well as in
`docker/distribution` packages

### CODE REVIEW

**Comment by cdrage** [comment]
@surajnarwade 
Please update commit description message (more detail, not just resolves X).

Tests would be good too!

**Comment by cdrage** [comment]
Hey @surajnarwade I don't see why there is an issue with this. You are only updating Version 3 of Docker Compose which uses `docker/cli`, you shouldn't be touching / using `docker/libcompose`.

I also see that `docker/cli` is already using the most up-to-date logrus here: https://github.com/docker/cli/commit/e3b7700fb9ac7df392e4915978bfc714d9dacad8

What issues are you running into? Please investigate and let me know!

**Comment by cdrage** [comment]
Also the error is just a simple renaming of `Sirupsen` in completion.go:

```
cmd/completion.go:9:2: cannot find package "github.com/Sirupsen/logrus" in any of:
	/home/travis/gopath/src/github.com/kubernetes/kompose/vendor/github.com/Sirupsen/logrus (vendor tree)
	/home/travis/.gimme/versions/go1.6.linux.amd64/src/github.com/Sirupsen/logrus (from $GOROOT)
	/home/travis/gopath/src/github.com/Sirupsen/logrus (from $GOPATH)
make: *** [test-unit-cover] Error 1
```

**Comment by cdrage** [comment]
After the rename it seems to work:

```
github.com/kubernetes/kompose  master ✔                                                                                                                                                                                                                                  22h19m  
▶ git-pull-pr 846
Upstream:  kubernetes/kompose
Already on 'master'
error: branch 'pr_846' not found.
remote: Counting objects: 102, done.
remote: Compressing objects: 100% (22/22), done.
remote: Total 102 (delta 55), reused 83 (delta 55), pack-reused 19
Receiving objects: 100% (102/102), 70.32 KiB | 0 bytes/s, done.
Resolving deltas: 100% (55/55), completed with 49 local objects.
From github.com:kubernetes/kompose
 * [new ref]         refs/pull/846/head -> pr_846
Switched to branch 'pr_846'

github.com/kubernetes/kompose  pr_846 ✔                                                                                                                                                                                                                                      2d  
▶ vim

github.com/kubernetes/kompose  pr_846 ✗                                                                                                                                                                                                                                    2d ⚑  
▶ make bin
go build -ldflags="-w -X github.com/kubernetes/kompose/cmd.GITCOMMIT=efb7552" -o kompose main.go

github.com/kubernetes/kompose  pr_846 ✗                                                                                                                                                                                                                                    2d ⚑  
▶ 
```

**Comment by surajnarwade** [comment]
@cdrage after renaming it works locally but not in CI
after investing I found out that, we have to update `openshift` from vendor, which may cause trouble

**Comment by cdrage** [comment]
Please troubleshoot and see if it works then.

File: glide.yaml [hunk 1]
```diff
@@ -24,15 +24,15 @@ import:
 
 # We use libcompose to parse v1 and v2 of Docker Compose
 - package: github.com/docker/libcompose
-  version: 4a647d664afbe05c41455c9d534d8239671eb46a
+  version: 57bd716502dcbe1799f026148016022b0f3b989c
```
Comment (cdrage) [inline glide.yaml]: don't update libcompose, you don't need to.

Comment (surajnarwade) [reply glide.yaml]: Sirupsen :(

**Comment by cdrage** [comment]
To give an update of what's happening after talking to @surajnarwade 

To update docker/cli we had to update vendoring. To update vendoring with the logrus errors, we had to update libcompose. To update libcompose, we had to update OpenShift vendoring due to conflicts, now we have to update and somehow migrate from 1.4 of OpenShift to 3.2 (I believe) in order to get all of this to work. :+1:

**Comment by surajnarwade** [comment]
@cdrage @kadel instead of updating openshift, I have replaced `Sirupsen` with `sirupsen` in `docker/distribution`

**Comment by cdrage** [comment]
@surajnarwade Can you please provide details on what you did?

Was it a PR that updated docker/distribution?

Did you manually edit the vendoring files?

Have you updated Makefile to sed replace Sirupsen to sirupsen, or anything else for that matter?

**Comment by surajnarwade** [comment]
@cdrage @kadel m To give brief idea about this PR,

I have Updated `docker/cli` to get latest changes that @cdrage pushed for build v3 support

these changes comes with `sirupsen` case collsion problem, so updated `docker/libcompose` as well

after fixing `sirupsen` case collision, it again occured with package `docker/distribution` which is dependency of `openshift/origin`, since updating `openshift/origin` is too much tedious task.


To tackle this situation, I updated `imports` from `kompose` manually and  `docker/distribution` (from vendor) by adding line in Makefile and Now kompose seems to be working fine :)

I also added this thing to Makefile with `make vendor-update`

**Comment by cdrage** [comment]
Awesome :+1: thanks for the detailed explanation!

One last thing, could you put the `make vendor-update` updates to Makefile into a separate commit so we can see what changes you've made? So in total, there should be three commits. One to the code, one to the Makefile and the other actually updating the vendoring files.

**Comment by surajnarwade** [comment]
@cdrage gotcha

**Comment by surajnarwade** [comment]
@cdrage now you can easily review it :)

**Comment by cdrage** [comment]
@surajnarwade Your makefile commit is messy: https://github.com/kubernetes/kompose/pull/846/commits/c4044b56a56a1b6c34db88b35686555e0b8b5195

Seems like you included some vendor (libcompose) files and there are some code that you're modifying in `pkg`...

So to go into detail:
Commit 1: Updates to Kedge code (pkg, renaming sirupsen, etc.)
Commit 2: Updates to Makefile (and that's it!)
Commit 3: Any updates to vendoring file.

**Comment by surajnarwade** [comment]
@cdrage , Updated commits, review needed

**Comment by cdrage** [comment]
@surajnarwade tests aren't passing  on travis :(

**Comment by surajnarwade** [comment]
@cdrage all green :)

File: script/test/cmd/tests.sh [hunk 2]
```diff
@@ -84,6 +84,11 @@ cmd="kompose --provider=openshift -f $KOMPOSE_ROOT/script/test/fixtures/nginx-no
 sed -e "s;%VERSION%;$version;g" -e "s;%CMD%;$cmd;g" -e "s;%URI%;$uri;g" -e "s;%REF%;$branch;g" $KOMPOSE_ROOT/script/test/fixtures/nginx-node-redis/output-os-template.json > /tmp/output-os.json
 convert::expect_success_and_warning "kompose --provider=openshift -f $KOMPOSE_ROOT/script/test/fixtures/nginx-node-redis/docker-compose.yml convert --stdout -j --build build-config" "/tmp/output-os.json" "$warning"
 
+# Replacing variables with current branch and uri
+# Test BuildConfig v3
+cmd="kompose --provider=openshift -f $KOMPOSE_ROOT/script/test/fixtures/nginx-node-redis/docker-compose-v3.yml convert --stdout -j --build build-config"
+sed -e "s;%VERSION%;$version;g" -e "s;%CMD%;$cmd;g" -e "s;%URI%;$uri;g" -e "s;%REF%;$branch;g" $KOMPOSE_ROOT/script/test/fixtures/nginx-node-redis/output-os-template-v3.json > /tmp/output-os.json
+convert::expect_success_and_warning "kompose --provider=openshift -f $KOMPOSE_ROOT/script/test/fixtures/nginx-node-redis/docker-compose-v3.yml convert --stdout -j --build build-config" "/tmp/output-os.json" "$warning"
```
Comment (cdrage) [inline script/test/cmd/tests.sh]: There should be another test added for Kubernetes.

**Comment by cdrage** [comment]
Code / implementation LGTM, works on my side perfectly.

After merging, can you please update the documentation (conversion.md document) as well as open an issue that we need to add a test for testing Kubernetes build.

**Comment by surajnarwade** [comment]
> Above works when I changed it to my username.

Thanks @cdrage  :)

Added test for kubernetes as well
Ready to merge :1st_place_medal:

**Comment by cdrage** [comment]
Phew. Thanks A LOT for your contribution @surajnarwade  that was a lot of trouble-shooting. Merging in!

**Comment by surajnarwade** [comment]
Thanks @cdrage

**Pull request merged**